lazysnap/src/main/java/com/google/accompanist/lazysnap/SnappingFlingBehavior.kt

```diff
@@ -0,0 +1,460 @@
+import androidx.compose.foundation.gestures.ScrollScope
+import androidx.compose.foundation.lazy.LazyListItemInfo
+import androidx.compose.foundation.lazy.LazyListLayoutInfo
+import androidx.compose.foundation.lazy.LazyListState
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.getValue
+import androidx.compose.runtime.mutableStateOf
+import androidx.compose.runtime.remember
+import androidx.compose.runtime.setValue
+import io.github.aakira.napier.DebugAntilog
+import io.github.aakira.napier.Napier
+import kotlin.math.abs
+import kotlin.math.absoluteValue
+import kotlin.math.max
+import kotlin.math.min
+import kotlin.math.roundToInt
+
+private const val DebugLog = false
+
+@RequiresOptIn(message = "Accompanist Lazy Snap is experimental. The API may be changed in the future.")
+@Retention(AnnotationRetention.BINARY)
+annotation class ExperimentalLazySnapApi
+
+/**
+ * Default values used for [SnappingFlingBehavior] & [rememberSnappingFlingBehavior].
+ */
+@ExperimentalLazySnapApi
+object SnappingFlingBehaviorDefaults {
+    /**
+     * TODO
+     */
+    val snapAnimationSpec: AnimationSpec<Float> = spring(stiffness = 400f)
+
+    val maximumFlingDistance: LazyListLayoutInfo.() -> Int = { Int.MAX_VALUE }
+}
+
+/**
+ * Create and remember a snapping [FlingBehavior] to be used with [LazyListState].
+ *
+ * @param lazyListState The [LazyListState] to update.
+ * @param decayAnimationSpec The decay animation spec to use for decayed flings.
+ * @param snapAnimationSpec The animation spec to use when snapping.
+ */
+@ExperimentalLazySnapApi
+@Composable
+fun rememberSnappingFlingBehavior(
+    lazyListState: LazyListState,
+    snapOffsetForItem: LazyListLayoutInfo.(LazyListItemInfo) -> Int = SnapOffsets.Center,
+    maximumFlingDistance: LazyListLayoutInfo.() -> Int = SnappingFlingBehaviorDefaults.maximumFlingDistance,
+    decayAnimationSpec: DecayAnimationSpec<Float> = rememberSplineBasedDecay(),
+    snapAnimationSpec: AnimationSpec<Float> = SnappingFlingBehaviorDefaults.snapAnimationSpec,
+): SnappingFlingBehavior = remember(
+    lazyListState,
+    snapOffsetForItem,
+    maximumFlingDistance,
+    decayAnimationSpec,
+    snapAnimationSpec
+) {
+    SnappingFlingBehavior(
+        lazyListState = lazyListState,
+        snapOffsetForItem = snapOffsetForItem,
+        maximumFlingDistance = maximumFlingDistance,
+        decayAnimationSpec = decayAnimationSpec,
+        snapAnimationSpec = snapAnimationSpec,
+    )
+}
+
+/**
+ * TODO
+ */
+@Suppress("unused")
+object SnapOffsets {
+    /**
+     * TODO
+     */
+    val Start: LazyListLayoutInfo.(LazyListItemInfo) -> Int = { 0 }
+
+    /**
+     * TODO
+     */
+    val Center: LazyListLayoutInfo.(LazyListItemInfo) -> Int = {
+        ((viewportEndOffset + viewportStartOffset) / 2) - (it.size / 2)
+    }
+
+    /**
+     * TODO
+     */
+    val End: LazyListLayoutInfo.(LazyListItemInfo) -> Int = { viewportEndOffset - it.size }
```

Comment:
this api is a bit confusing. how it was decided what is the receiver and what is the param? I would prefer both to be a param I think

Reply:
Moved `LazyListLayoutInfo` to be a param. Less confusing? My thought is that most apps will use one of these provided impls, but they can easily do something custom if they wish.

Reply:
yep, looks good